### PR TITLE
Simple Backup and Restore

### PR DESCRIPTION
**Pull request by matthme**
This PR is adds the backend functionality for a simple backup and restore mechanism:
- It adds a function that allows to backup all launcher data to a specified folder
- It adds a function to restore data from a folder of backup data

The idea is that the backup function may for example be called when the launcher is being quit and restoration from a backup could happen as part of the setup flow when starting with a new launcher from scratch.

This approach has amongst others the following limitations/drawbacks:
* It is heavy in data storage as it backs up almost all data
* The backed up data is not compressed into a single file as compression of this amount of data takes a long time and compression did not even seem to make the end-result considerably smaller
* The problem of the case where a device breaks while the launcher is running and the backup then only contains partial source chain data which could then lead to source chain forks when restoring from that backup is not solved

Backing up to, from and across different OS has been tested (Windows 10, macOS Catalina, Ubuntu 22.04) and seems to work fine.

There is no associated UI yet to actually use it. For testing I used a very rudimentary approach of manually writing to a variable on the window object which you can see in [this commit](https://github.com/holochain/launcher-electron/pull/166/commits/1266c53d37f70dba92c2455108c428a122d7dff4) where I removed the code I've been using to test it.

### CODE REVIEW

**Comment by matthme** [comment]
Side note - I think the practice that you started on scaffolding to get PR reviews from a broader circle of people is a good idea so I will experiment with adopting it here as well @c12i .

File: src/main/filesystem.ts [hunk 1]
```diff
@@ -1,15 +1,28 @@
...
+const KEYSTORE_DIRNAME = 'lair';
+const HOLOCHAIN_DIRNAME = 'holochain';
+const CONFIG_DIRNAME = 'config';
+const CONDUCTOR_ENV_DIRNAME = 'dbs';
+const UIS_DIRNAME = 'uis';
+const HAPPS_DIRNAME = 'happs';
+const APPS_DIRNAME = 'apps';
+
+const BACKUP_INFO_FILENAME = 'backup.info.json';
+const BACKUP_LOG_FILENAME = 'backups.log';
```
Comment (mrruby) [inline src/main/filesystem.ts]: can we move it to const.ts file?

**Review by mrruby — CHANGES_REQUESTED** [review]

File: src/main/filesystem.ts (hunk 1 above)
Comment (matthme) [reply src/main/filesystem.ts]: Okay, moved :+1:

Comment (mrruby) [reply src/main/filesystem.ts]: Thank you!

**Review by mrruby — APPROVED** [review]

File: src/main/filesystem.ts [hunk 2]
```diff
@@ -216,6 +253,318 @@ export class LauncherFileSystem {
...
+  get launcherConfigPath() {
+    return path.join(this.configDir, 'launcher-config.json');
+  }
+
+  get launcherConfig(): LauncherConfig | undefined {
+    if (fs.existsSync(this.launcherConfigPath)) {
+      const configStr = fs.readFileSync(this.launcherConfigPath, 'utf-8');
+      return JSON.parse(configStr) as LauncherConfig;
+    }
+    return undefined;
```
Comment (c12i) [inline src/main/filesystem.ts]:
```suggestion
    if (!fs.existsSync(this.launcherConfigPath)) return;
    const configStr = fs.readFileSync(this.launcherConfigPath, 'utf-8');
    return JSON.parse(configStr) as LauncherConfig;
```

File: src/main/filesystem.ts [hunk 3]
```diff
@@ -216,6 +253,318 @@ export class LauncherFileSystem {
...
+      return launcherConfig;
+    }
+  }
+
+  get backupLocation() {
+    const launcherConfig = this.launcherConfig;
+    if (launcherConfig) {
+      return launcherConfig.backupLocation;
+    }
+    return undefined;
```
Comment (c12i) [inline src/main/filesystem.ts]:
```suggestion
    if (!launcherConfig) return;
    return launcherConfig.backupLocation;
```

File: src/main/filesystem.ts [hunk 4]
```diff
@@ -216,6 +253,318 @@ export class LauncherFileSystem {
...
+      } else {
+        const errLog = `${platformLogString('BACKUP')} ERROR: Failed to copy keystore directory: ${e}`;
+        fs.appendFileSync(backupLogPath, errLog);
+        throw Error(`Failed to copy keystore directory: ${e}`);
+      }
+    }
+
+    // TODO check for .happ files and UIs that are not in use and don't copy them over
+
+    try {
```
Comment (c12i) [inline src/main/filesystem.ts:333]: Similar log as above `console.log('backing up lair...');` missing

File: src/main/filesystem.ts (hunk 2 above)
Comment (matthme) [reply src/main/filesystem.ts]: should be addressed now.

File: src/main/filesystem.ts (hunk 3 above)
Comment (matthme) [reply src/main/filesystem.ts]: addressed.

File: src/main/filesystem.ts (hunk 4 above)
Comment (matthme) [reply src/main/filesystem.ts:333]: added console log as well here

**Review by c12i — APPROVED** [review]

**Pull request merged**